Add map rasterization to improve client responsiveness when zoomed out.

RoosterDragon opened the pull request:
Fixes #31

Currently when zoomed out in the client with many boards visible, scrolling performance is poor. The large number of SVGs visible in the DOM tanks the browser's rendering performance.

To tackle this, introduce a new map-rasterization mode which caches the map image onto a canvas element. Introduce a policy that when zoomed out sufficiently, we remove the SVGs elements from the DOM and instead display the canvas element.

This greatly improves scrolling performance when zoomed out, as the browser has fewer DOM elements to deal with. The downside is that inputting orders is impossible until zooming back in Since users are likely to zoom in to give the required fidelity for clicking regions to input orders, this seems like a reasonable policy. But others policies are possible.

Code-wise there is a penalty in that we need to create parallel code paths to render elements to a canvas, as well as generating elements for React.

----

This is quite heavy in terms of extra code required, but I'll provide it in case anybody wants to mess with it or can otherwise streamline it.

Test case: Open any large game and zoom out. With this PR, once the map images have loaded, you can smoothly pan the map around. Previously it would be laggy. Here's an example:

https://github.com/user-attachments/assets/f3e8efe2-37db-43eb-8809-b80227d3bf94

`rasterizeDisplayFallback = false` means you'll be able to see the maps load in when zoomed out. You can also set `rasterizeDisplayFallback = true` which means we'll display the SVGs instead until the maps load in. That provides a more seamless visual experience whilst loading is happening, and you'll still get the performance benefits once everything has finished loading.

Example with the maps partially loaded:
![image](https://github.com/user-attachments/assets/844b3f3f-7b79-49e7-8575-7cbde7ec9dc9)

Now let's look at the comparison, here we are at zoom just above 0.6:
![image](https://github.com/user-attachments/assets/cdb37c27-b36d-40f4-a649-1ca8f1b7fa6e)

Then just under 0.6, at which point the image is used instead:
![image](https://github.com/user-attachments/assets/c52b58cf-bf8e-4bff-989f-2290305f07cc)

`rasterizeScaleThreshold` can be used to change where this transition occurs. The `rasterizeSize` settings controls the image quality of these snapshots, you can render to larger images for better quality, but it costs more memory to keep around.

MichaelEbert commented on client/src/components/world/boards/BoardLayer.tsx:
```diff
@@ -7,10 +7,36 @@ import WorldContext from '../../context/WorldContext';
+
+  // Rasterization shares a queue, this allows one item to be resolved at a time and displayed incrementally for better UX.
+  // Without the queue, we'd try and rasterize everything at once which chugs the browser for a while.
+  const rasterizeQueue = useMemo<WorkQueue>(() => {
```
should this be in WorkQueueContext instead of BoardLayer? seems like queue management

MichaelEbert commented on client/public/assets/icons/ArmyIcon.svg:
Does this duplicate all the SVGs in src/assets?
can we put them in one place or the other?

Oliveriver replied:
I'm also confused here. If the move to the public folder is necessary for importing the rasterised images, then the assets folder in src should be removed and imports updated.

RoosterDragon replied:
I did try updating the imports, e.g. `import ADR from '../../public/assets/map/ADR.svg?react';`, but this causes warnings from `yarn dev`

```
Assets in public directory cannot be imported from JavaScript.
If you intend to import that asset, put the file in the src directory, and use /src/assets/map/ADR.svg?react instead of /public/assets/map/ADR.svg?react.
If you intend to use the URL of that asset, use /assets/map/ADR.svg?url&react.
```

So duplicating everything was the easy way out.

I'd love to remove the duplication too, but I'll need a steer on an approach that works.

Oliveriver replied:
I still feel like duplicating every SVG file is quite a big problem so I'm not sure I can accept this PR until that's resolved. I'll have a quick look into it tonight.

RoosterDragon replied:
After some research I found two helpful things to bring this all together:

- [Importing Asset as String](https://vite.dev/guide/assets#importing-asset-as-string) - `?raw` will allow access to the stringified content.
- [Glob Import](https://vite.dev/guide/features#glob-import) - Allows import of all the maps via glob, so I don't have to duplicate a bunch of imports by hand for both `?react` and `?raw` queries.

Everything now remains in the assets folder and nothing is added to the publiuc folder.

Oliveriver commented on client/src/components/user-interface/InputModeButtonList.tsx:
```diff
@@ -9,10 +12,40 @@ const InputModeButtonList = () => {
+    return () => {
+      onChangeCallbacks.delete(callback);
+    };
+  }, [onChangeCallbacks, scale]);
```
Is there a way to get around replacing the callback every time `scale` changes? I remember I had a go at something similar and ended up with effectively the same `useEffect` and that bugged me. Maybe using the alternate version of `setScale`, i.e. `setScale(currentScale => ...)`, could do something.

RoosterDragon replied:
I have no idea to be honest - I just copied your working version and left it at that. :)

Oliveriver commented on client/src/components/user-interface/InputModeButtonList.tsx:
```diff
@@ -9,10 +12,40 @@ const InputModeButtonList = () => {
-      {inputButtons}
-    </div>
+    <>
+      {showRasterized && (
```
I feel like it's clearer to the user and just looks better to hide the input mode buttons when orders can't be entered, rather than showing this message.

Oliveriver commented on client/src/components/user-interface/InputModeButtonList.tsx:
```diff
@@ -9,10 +12,40 @@ const InputModeButtonList = () => {
+
+  const [scale, setScale] = useState(initialScale);
+
+  useEffect(() => {
```
I understand why this is in `InputModeButtonList`, but it feels like a separate concern. Maybe it should be moved to the context (a bit like MichaelEbert's suggestion below) and then this component can use the context.

RoosterDragon replied:
Created a ScaleContext.

Oliveriver commented on client/src/components/world/boards/Map.tsx:
```diff
@@ -1,74 +1,169 @@
+
+  const [scale, setScale] = useState(initialScale);
+
+  useEffect(() => {
```
Moving the `useEffect` above to a context could also remove the need for this duplication. This feels quite expensive as it is, especially since this will apply for every map.

RoosterDragon replied:
Created a ScaleContext.

Oliveriver commented on client/src/components/world/boards/Map.tsx:
```diff
@@ -1,74 +1,169 @@
+      let regionsLeft = regionProps.length;
+      Object.values(regionProps).forEach((regionProp) => {
+        RasterizeRegion(regionProp, canvas, context, () => {
+          if (--regionsLeft === 0) {
```
This syntax broke my brain for a second. I'd prefer the decrement and comparison to be split over separate lines.

Oliveriver commented on client/src/components/world/boards/Map.tsx:
```diff
@@ -1,74 +1,169 @@
+      {showRasterized && (rasterized || !rasterizeDisplayFallback)
+        ? null // Remove SVGs from DOM, to improve responsiveness when zoomed out.
+        : getRegions(board, isShowingCoasts, world, currentMode, currentOrder).map((e) => (
           <Region
```
Can just use spread syntax for the component, i.e. `{...e}`. Also I'd like a more descriptive name than `e`.

Oliveriver commented on client/src/components/context/WorkQueueContext.tsx:
```diff
@@ -0,0 +1,8 @@
+const noopQueue = { push: () => {} };
+const WorkQueueContext = createContext<WorkQueue>(noopQueue);
+
+export type WorkQueue = { push: (workItem: (oncomplete: () => void) => void) => void };
```
Should be `onComplete`. Same in a few other places.

Oliveriver commented on client/src/components/world/boards/Region.tsx:
```diff
@@ -36,6 +37,87 @@ type RegionProps = {
   isVisible?: boolean;
 };
 
+export const RasterizeRegion = (
```
This is not a component, so shouldn't be capitalised. It might even be worth putting in a separate file.

RoosterDragon replied:
Adjusted casing.

If you'd like it in a separate file, I'm open to that. I've initially put them in the same file, since the rasterization pass and React rendering are obviously tightly coupled - if you update what React renders you'll need to update the rasterisation pass to match. So having them live in the same place to avoid forgetting seemed best.

Oliveriver replied:
True, I think it makes sense to keep them in the same files as the components then.

Oliveriver commented on client/src/components/world/boards/Region.tsx:
```diff
@@ -36,6 +37,87 @@ type RegionProps = {
+      }
+    };
+
+    axios.get<string>(`assets/map/${file}.svg`).then((r) => {
```
Well this is odd. I've never seen axios used to get local files. I guess it makes sense though...?

I'm generally not a fan of `then` syntax though, so I'd prefer it if this was changed to use `await`.

Oliveriver commented on client/src/components/world/boards/UnitIcon.tsx:
```diff
@@ -11,6 +12,62 @@ type UnitIconProps = {
   variant?: 'world' | 'overlay';
 };
 
+export const RasterizeUnitIcon = (
```
Similar to Region. This is not a component, and argubaly should go elsewhere.

RoosterDragon replied:
Casing fixed.

Oliveriver commented on client/src/utils/constants.ts:
```diff
@@ -17,6 +17,15 @@ export const orderArrowStartSeparation = 20;
+// Size of the generated image. Larger sizes improve quality, smaller reduces memory usage.
+export const rasterizeSize = 300;
+// At what zoom level under which we display the images instead of the SVGs.
+export const rasterizeScaleThreshold = 0.6;
```
My gut feeling is this should be lowered a fair bit. I tried it out and found it frustrating to have to zoom in and out so much to enter orders between boards. I felt comfortable at about 0.2.

RoosterDragon replied:
Changed to 0.2, and the rasterizeSize to 100 accordingly.

Oliveriver reviewed (changes requested):
Impressive work, although I'm worried about the different paradigms that will be in play together here.

Also, very nitpicky, but I've stuck to British English where possible in this repo, so I'm annoyingly going to ask for this to use "rasterise" instead of "rasterize".

Oliveriver commented on client/src/utils/constants.ts:
```diff
@@ -17,6 +17,15 @@ export const orderArrowStartSeparation = 20;
+// Enable showing maps as rasterised images instead of SVGs. Improves UI responsiveness when zoomed out, but disables interactivity.
+export const rasteriseEnabled = true;
+// Size of the generated image. Larger sizes improve quality, smaller reduces memory usage.
+export const rasteriseSize = 100;
```
People like to take 8K screenshots of large games and I've realised this change doesn't mesh well with that with a small raster size. I'd suggest increasing this to 200 at least.

RoosterDragon replied:
To make this easier to manage, I've swapped out this constant for a `rasteriseFactor = 1`. 1 will generate a full size image, 0.5 would be half size, etc.

RoosterDragon commented:
Rebased to account for #50, and changed the work queue to prioritize rendering maps near the centre of the viewport first. This means we're more likely to have the maps we want to display as images ready when the users zooms out.

RoosterDragon commented:
> I'm not sure how, but I'm getting much worse performance with the latest version of this branch. Ironically everything is super fast when zoomed out and quite jittery when zoomed in.

I'll have to poke around, wondering if #50 could also be contributing.

RoosterDragon commented:
Ok discovered the cause of the slowdowns:
- Generating all the images caused a throughly unreasonable amount of memory traffic (10 GBs or so) that would then lock up the browser for a bit whilst it cleaned all that up - solved by adding an image cache so we can reuse images across multiple maps rather than having to load them fresh every time. This also improves the initial load time for the rasterised versions considerably.
- Panning was causing map to re-render which is expensive because this in turn renders all the regions. Applied a `useMemo` to fix this.

For me, that restores the lost performance when zoomed in.

Oliveriver reviewed (approved):
Excellent work! There's still some jankiness at certain scales when panning quickly but it's much more manageable.
